core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -961,11 +961,18 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
   }
 
   /** Build the union of a list of RDDs. */
-  def union[T: ClassTag](rdds: Seq[RDD[T]]): RDD[T] = new UnionRDD(this, rdds)
+  def union[T: ClassTag](rdds: Seq[RDD[T]]): RDD[T] = {
+    val partitioners = rdds.map(_.partitioner).toSet
```

Comment:
If `_.partitioner` is an option, then I think this can be simplified by using `flatMap` instead of `map`, since that would just let you check whether `partitioners.size == 1` on the next line without having to have the `isDefined` check as well.